### ISSUE TITLE
Infinite recursion on save("test.pgm")

### ISSUE THREAD

**Issue by mewo2** [JuliaIO/FileIO.jl#53]
I'm not sure if this is a bug in FileIO, or in the underlying Netpbm (or even a bug at all - I only noticed it because of a misplaced parenthesis). Anyway, one can induce an infinite recursion with the following code:
```julia
using Images
save("test.pgm")
```
It seems that in the absence of any data to save, the call at `loadsave.jl` line 92 is recursing infinitely.

**Comment by SimonDanisch** [comment]
It's a bug! ;) It should be fixed at some point!

**Comment by staticfloat** [comment]
What is the proper way to avoid this?  I'm trying to enforce the usage of certain parameters when `save()`'ing out a `.flac` file, but if I enforce a certain method signature, then we get recursive compilation which kills Julia.  :(

The best I can come up with is defining all my `save()` methods as `save(f::File{format"FLAC"}, data..., kwargs...)` so as to catch _all_ incoming method signatures, and then manually inspecting `data` and `kwargs`, which is kind of a code smell, in my opinion.  I think we should have some way of enforcing which module the `save()` function is being pulled from in `loadsave.jl` on line 92, so that we can actually fail to compile without hitting infinite recursion.

**Comment by SimonDanisch** [comment]
Just to make sure that I understand correctly:
You have a certain signature with different kw_args/data, which is only supported by a certain Module, so you'd like to enforce using that Module?

**Comment by staticfloat** [comment]
My example is that I defined `save(f::File{format"FLAC"},  data, samplerate)`, in an effort to force the user to provide a samplerate along with their filename and data in order to save a `.flac` file.  However, if the user calls `save("test.flac", [1,2,3])` the compiler wigs out because [this line](https://github.com/JuliaIO/FileIO.jl/blob/e10242de106c8c27823fb920fad787beef5aff26/src/loadsave.jl#L95) recurses infinitely.  This is because `FLAC` imports `FileIO` to overload `save()`, and so that line just ends up calling itself.

**Comment by SimonDanisch** [comment]
Okay, got it!
Not really solving the problem, but would it make sense to make the sample rate part of the data?

**Comment by SimonDanisch** [comment]
Okay so you shouldn't import FileIO.save/load into FLAC.jl...
And then in line 95, we should add:
```Julia
Library.save == FileIO.save && error("Library does not have a save method")
```
Can you confirm, that this solves it?

**Comment by staticfloat** [comment]
Oh, duh.  Just don't import them since I don't need them.  Sigh.  Thank you.  I'm obviously way too tired for this.  I can confirm your fix above also plugs the hole in the event of confused users like myself.  ;)

May I suggest you change the wording to `does not have a matching save method`, to help guide package authors a bit more forcefully.

**Comment by SimonDanisch** [comment]
Yeah that's probably not very clear in the documentation and also a weird design, I must admit...

Oh definitely! That was just a placeholder ;)

**Comment by timholy** [comment]
Closed by #79
